File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean find(TreeNode root, TreeNode node, int target) {
        if(root == null) return false;
        if(root.val < target) return find(root.right, node, target);
        if(root.val > target) return find(root.left, node, target);
        if(root == node) return false;
        if(root.val == target) return true;
        return false;
    }
    boolean dfs(TreeNode root, TreeNode node, int target) {
        if(node == null) return false;
        if(find(root, node, target - node.val)) return true;
        if(dfs(root, node.left, target)) return true;
        if(dfs(root, node.right, target)) return true;
        return false;
    }
    public boolean findTarget(TreeNode root, int k) {
        return dfs(root, root, k);
    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find(self, root: Optional[TreeNode], node: Optional[TreeNode], target: int) -> bool:
        if not root: return False
        if root.val < target: return self.find(root.right, node, target)
        if root.val > target: return self.find(root.left, node, target)
        if root == node: return False
        if root.val == target: return True
        return False
    def dfs(self, root: Optional[TreeNode], node: Optional[TreeNode], target: int) -> bool:
        if not node: return False
        if self.dfs(root, node.left, target): return True
        if self.dfs(root, node.right, target): return True
        if self.find(root, node, target - node.val): return True
        return False
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        return self.dfs(root, root, k)